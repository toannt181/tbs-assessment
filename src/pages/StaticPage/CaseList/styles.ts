import styled from 'styled-components'

export const CaseListContainer = styled.div``

export const Case = styled.li`
  margin: 16px;
  background: #fff;
  box-shadow: 4px 4px 16px #f1f1f1;
  border-radius: 4px;
  transition: 0.25s;
  flex-basis: 100%;
  padding: 16px;
  display: block;

  &:hover {
    box-shadow: 6px 6px 24px #e3e3e3;
  }
`

export const Label = styled.div`
  font-weight: bold;
  display: inline-block;
  margin-right: 16px;
`
