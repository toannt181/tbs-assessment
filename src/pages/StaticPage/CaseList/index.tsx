import React, { memo, useRef, useEffect } from 'react'
import { CaseListContainer, Case, Label } from './styles'
import { ICase } from 'types/ICase'

interface ICaseListProps {
  cases: ICase[]
  onReachBottom: () => void
}

const CaseList = ({ cases, onReachBottom }: ICaseListProps) => {
  const containerRef = useRef<HTMLDivElement>(null)
  const loadingRef = useRef<HTMLDivElement>(null)

  useEffect(() => {
    if (!loadingRef.current || !containerRef.current) return
    const loadingDom = loadingRef.current as Element

    const callback = (entries: IntersectionObserverEntry[]) => {
      entries.forEach((entry) => {
        if (entry.isIntersecting) {
          onReachBottom()
        }
      })
    }
    const observer = new IntersectionObserver(callback)

    observer.observe(loadingDom)

    return () => observer.unobserve(loadingDom)
  }, [onReachBottom])

  return (
    <CaseListContainer ref={containerRef}>
      <ul data-testid="list">
        {cases.map((item) => (
          <Case key={item.id}>
            <div>
              <Label>Country</Label>
              {item.Country}
            </div>
            <div>
              <Label>CountryCode</Label>
              {item.CountryCode}
            </div>
            <div>
              <Label>Province</Label>
              {item.Province}
            </div>
            <div>
              <Label>Latitude</Label>
              {item.Lat}
            </div>
            <div>
              <Label>Longitude</Label>
              {item.Lon}
            </div>
            <div>
              <Label>Date</Label>
              {item.Date}
            </div>
          </Case>
        ))}
      </ul>
      <div ref={loadingRef} />
    </CaseListContainer>
  )
}

export default memo(CaseList)
