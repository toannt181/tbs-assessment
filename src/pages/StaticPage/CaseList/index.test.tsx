import CaseList from '.'
import React from 'react'
import { render, screen } from '@testing-library/react'
import mock from './mock'

global.IntersectionObserver = mock

describe('CaseList Component', () => {
  it('should render the component', () => {
    const props = {
      cases: [],
      onReachBottom: () => {},
    }
    const { container } = render(<CaseList {...props} />)
    expect(container).toMatchSnapshot()
  })

  it('should render the list with items', () => {
    const props = {
      cases: [
        {
          id: 1,
          CountryCode: 'CountryCode',
          Country: 'Country',
          Province: 'Province',
          Lat: 'Lat',
          Lon: 'Lon',
          Date: 'Date',
        },
      ],
      onReachBottom: () => {},
    }
    render(<CaseList {...props} />)
    expect(screen.getByTestId('list').children).toHaveLength(1)
  })
})
