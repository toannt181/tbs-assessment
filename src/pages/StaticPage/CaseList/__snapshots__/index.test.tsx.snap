// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CaseList Component should render the component 1`] = `
<div>
  <div
    class="sc-bdfBwQ"
  >
    <ul
      data-testid="list"
    />
    <div />
  </div>
</div>
`;
